File: LeetCode/Medium/210. Course Schedule II/Course Schedule II.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        # 인접리스트, 진입차수 초기화
        N = numCourses
        graph = [[] for _ in range(N)]
        ind = [0] * N

        for v, u in prerequisites:
            graph[u].append(v)
            ind[v] += 1

        # 큐 초기화
        q = deque([])
        result = []

        for i in range(N):
            if ind[i] == 0:
                q.append(i)

        # 위상정렬
        while q:
            u = q.popleft()
            result.append(u)

            for v in graph[u]:
                ind[v] -= 1
                if ind[v] == 0:
                    q.append(v)

        if len(result) != N:
            result = []

        return result

File: LeetCode/Medium/210. Course Schedule II/Course Schedule II.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        // 인접리스트, 진입차수 초기화
        int N = numCourses;
        List<List<Integer>> graph = new ArrayList<>();
        int[] inDegrees = new int[N];
        
        for (int i = 0; i < N; i++)
            graph.add(new ArrayList<>());

        for (int[] parts : prerequisites) {
            int v = parts[0];
            int u = parts[1];

            graph.get(u).add(v);
            inDegrees[v]++;
        }

        // 큐 초기화
        Queue<Integer> q = new LinkedList<>();
        int[] result = new int[N];
        int idx = 0;

        for (int i = 0; i < N; i++) {
            if (inDegrees[i] == 0)
                q.offer(i);
        }

        // 위상 정렬
        while (!q.isEmpty()) {
            int u = q.poll();
            result[idx++] = u;

            for (int v : graph.get(u)) {
                inDegrees[v]--;
                if (inDegrees[v] == 0)
                    q.offer(v);
            }
        }

        // 사이클 확인
        if (idx != N)
            return new int[]{};

        return result;
    }
}